### PR TITLE
#937 Add option for AuthMe to run in sync

### PR DESCRIPTION
**Pull request by ljacqu**
Please review these changes for #937

- Added new setting to define whether tasks should be async (default) or sync
- Added new methods in BukkitService to schedule tasks async / sync based on that setting
  * If it's sync we just run the code directly, it's not scheduled as a sync delayed task. **Is that OK?**

I don't understand when we us BukkitScheduler's `runTask()` vs. `scheduledSyncDelayedTask()`. We use both to run sync code inside an async thread, e.g. in AsynchronousJoin. Does anyone know the difference?

Finally, the change to the Initializer is ugly but it can be refactored once https://github.com/ljacqu/DependencyInjector/issues/32 has been implemented.

### CODE REVIEW

**Comment by sgdc3** [comment]
> I don't understand when we us BukkitScheduler's runTask() vs. scheduledSyncDelayedTask(). We use both to run sync code inside an async thread, e.g. in AsynchronousJoin. Does anyone know the difference?

Both methods should not be used. The right way to schedule/run sync/async stuff is to use BukkitRunnable instead of Runnable

**Comment by sgdc3** [comment]
>If it's sync we just run the code directly, it's not scheduled as a sync delayed task. Is that OK?

I really don't know, we need to test this.

**Comment by ljacqu** [comment]
@sgdc3 Thanks for your comment. Regarding your first comment, if you look at the javadocs https://hub.spigotmc.org/javadocs/spigot/org/bukkit/scheduler/BukkitScheduler.html only the methods taking a BukkitRunnable are deprecated. I would wager that passing a Runnable is still fine or they'd be marked as deprecated, too. I still don't understand the difference between [runTask(Plugin, Runnable)](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/scheduler/BukkitScheduler.html#runTask(org.bukkit.plugin.Plugin,%20java.lang.Runnable)) and [scheduleSyncDelayedTask(Plugin, Runnable)](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/scheduler/BukkitScheduler.html#scheduleSyncDelayedTask(org.bukkit.plugin.Plugin,%20java.lang.Runnable)). The latter explicitly mentions that it will run on the main thread, maybe that? But I thought sync was always on the main thread?

As for your second comment, I tested locally and it worked as expected. But of course this doesn't tell us whether it's a good approach with 20+ players, for example. I did find a weird problem when unregistering and registering again—working on fixing it atm.

File: src/main/java/fr/xephi/authme/process/register/AsyncRegister.java [hunk 1]
```diff
@@ -172,7 +170,12 @@ private void passwordRegister(final Player player, String password, boolean auto
         }
 
         if (!service.getProperty(RegistrationSettings.FORCE_LOGIN_AFTER_REGISTER) && autoLogin) {
-            bukkitService.runTaskOptionallyAsync(() -> asynchronousLogin.login(player, "dontneed", true));
+            if (service.getProperty(PluginSettings.USE_ASYNC_TASKS)) {
+                bukkitService.runTaskAsynchronously(() -> asynchronousLogin.login(player, "dontneed", true));
+            } else {
+                bukkitService.scheduleSyncDelayedTask(
+                    () -> asynchronousLogin.login(player, "dontneed", true), SYNC_LOGIN_DELAY);
+            }
```
Comment (ljacqu) [inline src/main/java/fr/xephi/authme/process/register/AsyncRegister.java]: Not happy about this, but at this late hour I don't have a better idea.

**Comment by ljacqu** [comment]
Fixed the issue and played around more; everything as expected.

**Comment by sgdc3** [comment]
Please take a look there: http://wiki.bukkit.org/Scheduler_Programming

**Comment by ljacqu** [comment]
@sgdc3 I read the entire page, what's the takeaway?

**Comment by sgdc3** [comment]
Look at the BukkitRunnable section, the last phrase.

**Comment by ljacqu** [comment]
> Plugins should **prefer** defining a BukkitRunnable and calling the appropriate run method over directly scheduling a Runnable with the BukkitScheduler. 

I saw that, and I'm assuming it should be preferred as it gives more control e.g. in case we want to repeat some existing task. The bottom line is the verb "prefer." The methods aren't deprecated. The page states clearly not to pass BukkitRunnable to the BukkitScheduler, which corresponds to the deprecated methods. If you want to switch all of our uses of Runnable with BukkitRunnable, feel free to do so. :wink:

**Comment by ljacqu** [comment]
From Spigot IRC

> [00:59] == ljacqu [webchat@SpigotMC-ukvmt3.fe1m.0b6b.1205.2a02.IP] has joined #spigot
[00:59] -fry.spi.gt:#spigot- Replaying up to 10 lines of pre-join history spanning up to 60 seconds
[01:01] <ljacqu> Hi, quick question, the Wiki page on the Scheduler says BukkitRunnable over plain Runnable classes. Does anyone know the reason?
[01:01] <ljacqu> says BukkitRunnable \*should be preferred\* over Runnable classes
[01:02] == Chained_ [~Chained@SpigotMC-apppum.dip0.t-ipconnect.de] has joined #spigot
[01:03] <electronicboy> BukkitRunnables is the magical replacement to having to access the scheduler as well as adding a few extra feature; e.g. the ability to easily cancel a task
[01:04] <ljacqu> In other words, if I want to use Runnable (from an architectural perspective), I can do so without any kind of performance/technical disadvantage?
[01:04] <electronicboy> Pretty much, However there is very little reason to do so
[01:05] <ljacqu> Cool, thanks. Personally not a huge fan of a thread scheduling itself
[01:05] <Zee> Runnables get wrapped anyways

--> Canceling is still a fair point, though. This might be worth considering for the tasks we have for login reminders / antibot if they aren't BukkitRunnable already. For tasks like AsyncRegister we just "fire and forget" anyway.

**Pull request merged**